### PR TITLE
Fix PathFollow3D update issues

### PR DESCRIPTION
**Pull request by xiongyaohua**
This PR fix the following minor issues on PathFollow3D.

- transform is not updated after setting new flags such as `use_model_front`
- transform is not updated when the parent Path3D changes
- correct_posture() behaves differently when assuming model front
- _update_transform() was in immediate mode, could leads to chained calls on scene instantiation.

## Before:
https://github.com/godotengine/godot/assets/6020486/ffd7242e-73e7-41b5-97c9-ffbb3657da11

## After:
https://github.com/godotengine/godot/assets/6020486/f41d96c6-cd8e-484b-b590-5b375109bf2d

### CODE REVIEW

**Comment by YuriSizov** [comment]
Could you provide a test project that you've used to verify the fixes? Also, could you please amend the first line of the commit message to be a bit more explicit (the title of the PR is fine for that).

**Comment by xiongyaohua** [comment]
> Could you provide a test project that you've used to verify the fixes? 

I found the problems when try to revamp this PR https://github.com/godotengine/godot/pull/68873

The test file uploaded by @Calinou at the end is used.

> Also, could you please amend the first line of the commit message to be a bit more explicit (the title of the PR is fine for that).

Sure, changed.

**Comment by xiongyaohua** [comment]
@YuriSizov Hi, any chance to have this PR reviewed recently? It is relatively simple, and will greatly improve the user experience of this PR https://github.com/godotengine/godot/pull/80329

File: scene/3d/path_3d.cpp [hunk 1]
```diff
@@ -192,23 +194,25 @@ void PathFollow3D::_update_transform(bool p_update_xyz_rot) {
...
 		if (tilt_enabled) {
 			const real_t tilt = c->sample_baked_tilt(progress);
 
-			const Basis twist(forward, tilt);
+			const Basis twist(tangent, tilt);
 			t.basis = twist * t.basis;
 		}
 	}
 
+	// apply offset and scale
```
Comment (akien-mga) [inline scene/3d/path_3d.cpp:214]:
```suggestion
	// Apply offset and scale.
```

File: scene/3d/path_3d.cpp [hunk 2]
```diff
@@ -192,23 +194,25 @@ void PathFollow3D::_update_transform(bool p_update_xyz_rot) {
 		t.origin = pos;
 	} else {
 		t = c->sample_baked_with_rotation(progress, cubic, false);
+		Vector3 tangent = -t.basis.get_column(2); // Retain tangent for applying tilt.
+		t = PathFollow3D::correct_posture(t, rotation_mode);
+
+		// Switch z+ and z- if necessary.
```
Comment (akien-mga) [inline scene/3d/path_3d.cpp:200]:
```suggestion
		// Switch Z+ and Z- if necessary.
```

File: scene/3d/path_3d.cpp [hunk 3]
```diff
@@ -431,6 +435,7 @@ bool PathFollow3D::is_using_model_front() const {
 
 void PathFollow3D::set_loop(bool p_loop) {
 	loop = p_loop;
+	_update_transform();
```
Comment (akien-mga) [inline scene/3d/path_3d.cpp:438]: Will all these `_update_transform()` calls lead to doing multiple chained updates unnecessarily, e.g. when setting all the properties when constructing the object from a scene file?

Should this be cached with a dirty update pattern, or deferred?

Comment (xiongyaohua) [reply scene/3d/path_3d.cpp:438]: > Will all these _update_transform() calls lead to doing multiple chained updates unnecessarily

I am not sure, I just followed the existing convention, such as
```
void PathFollow3D::set_v_offset(real_t p_v_offset) {
	v_offset = p_v_offset;
	if (path) {
		_update_transform();
	}
}
```

> Should this be cached with a dirty update pattern, or deferred?

I am not sure it is worthwhile for two reasons
- `_update_transform` is not that expensive nor frequently called
- to defer the call we need to bind `_update_transform`, which should be kept away from end user.

Of course, I am not that familiar with the code base. If you think it need to be cached I can do it.

Comment (xiongyaohua) [reply scene/3d/path_3d.cpp:452]: > cached with a dirty update pattern

done

File: scene/3d/path_3d.cpp [hunk 4]
```diff
@@ -328,6 +344,7 @@ void PathFollow3D::_bind_methods() {
 
 	ClassDB::bind_method(D_METHOD("set_tilt_enabled", "enabled"), &PathFollow3D::set_tilt_enabled);
 	ClassDB::bind_method(D_METHOD("is_tilt_enabled"), &PathFollow3D::is_tilt_enabled);
+	ClassDB::bind_method(D_METHOD("_update_transform"), &PathFollow3D::_update_transform);
```
Comment (AThousandShips) [inline scene/3d/path_3d.cpp:347]: It is not necessary to bind to be able to call this deferred, you can use `callable_mp` AFAIK

Comment (xiongyaohua) [reply scene/3d/path_3d.cpp]: Right, it's also feels more straightforward. Thanks.

File: scene/3d/path_3d.h [hunk 5]
```diff
@@ -64,6 +64,8 @@ class PathFollow3D : public Node3D {
 	GDCLASS(PathFollow3D, Node3D);
 
 public:
+	void _update_transform(bool p_immediate = false);
```
Comment (YuriSizov) [inline scene/3d/path_3d.h:67]: If it's a public method, it shouldn't start with an underscore. You also should probably move it below where all other public methods are.

File: scene/3d/path_3d.cpp [hunk 6]
```diff
@@ -133,6 +133,8 @@ void Path3D::_curve_changed() {
 			PathFollow3D *child = Object::cast_to<PathFollow3D>(get_child(i));
 			if (child) {
 				child->update_configuration_warnings();
+				// Also update the transform of the PathFollow3Ds.
```
Comment (YuriSizov) [inline scene/3d/path_3d.cpp:136]: Comment is probably unnecessary here, code is simple enough.

File: scene/3d/path_3d.cpp [hunk 7]
```diff
@@ -170,7 +172,18 @@ void Path3D::_bind_methods() {
...
 //////////////
 
-void PathFollow3D::_update_transform(bool p_update_xyz_rot) {
+void PathFollow3D::_update_transform(bool p_immediate) {
+	// Schedule update later.
+	if (!p_immediate) {
+		transform_dirty = true;
+		callable_mp(this, &PathFollow3D::_update_transform).call_deferred(true);
+		return;
+	}
```
Comment (YuriSizov) [inline scene/3d/path_3d.cpp:181]: There should probably be 2 methods, one public with the `p_immediate` argument, and one private that is actually called if `p_immediate` is true or via this deferred call here.

File: scene/3d/path_3d.cpp [hunk 8]
```diff
@@ -185,30 +198,34 @@ void PathFollow3D::_update_transform(bool p_update_xyz_rot) {
 		return;
 	}
 
+	// Actual update.
+	transform_dirty = false;
```
Comment (YuriSizov) [inline scene/3d/path_3d.cpp:202]: It's probably safe to set it to `false` immediately after checking. The rest of the checks here ensure we are not in a state where we care about transforms anyway. And if this changes, then something somewhere should already call `update_transform` to force an update.

Setting `path` and path's curve does that. Baked length is an exception, it seems, but that may actually be a separate problem that also needs solving at some point.

**Comment by YuriSizov** [comment]
Other than what I mentioned above code looks fine and changes seem sensible.

@aaronfranke @kleonc Could you check the math, do these changes to transform updates make sense?

File: scene/3d/path_3d.h (hunk 5 above)
Comment (xiongyaohua) [reply scene/3d/path_3d.h]: fixed

File: scene/3d/path_3d.cpp (hunk 6 above)
Comment (xiongyaohua) [reply scene/3d/path_3d.cpp]: fixed

File: scene/3d/path_3d.cpp (hunk 7 above)
Comment (xiongyaohua) [reply scene/3d/path_3d.cpp]: fixed

File: scene/3d/path_3d.cpp (hunk 8 above)
Comment (xiongyaohua) [reply scene/3d/path_3d.cpp]: fixed.

**Review by akien-mga — APPROVED** [review]
Looks good to me.

**Comment by akien-mga** [comment]
Thanks!

**Comment by kleonc** [comment]
Note this should not be cherry-picked to 4.2 _on its own_ because of a regression: #95612.

**Comment by Steinbeuge** [comment]
I tested paths on 4.4 dev3 as at this time I can't use 4.3, there's still and issue with initial positions not being updated even when .progress or .progress_ratio is triggered with a zero value
https://github.com/godotengine/godot/issues/98602

**Pull request merged**